lib/sqlalchemy/ext/mypy/infer.py

```diff
@@ -413,9 +414,11 @@ def _extract_python_type_from_typeengine(
             n = api.lookup_fully_qualified("builtins.str")
             return Instance(n.node, [])
 
-    for mr in node.mro:
-        if mr.bases:
-            for base_ in mr.bases:
-                if base_.type.fullname == "sqlalchemy.sql.type_api.TypeEngine":
-                    return base_.args[-1]
-    assert False, "could not extract Python type from node: %s" % node
+    assert node.has_base("sqlalchemy.sql.type_api.TypeEngine"), (
+        "could not extract Python type from node: %s" % node
+    )
+    type_engine = map_instance_to_supertype(
```

Comment:
wow what is it?

Reply:
It takes an instance and a `TypeInfo` for a superclass with a generic parameter and figures out what the instance has filled in for the generic parameter. I did a lot of digging in the mypy internals to find this :)